### PR TITLE
Route-local fault filter config

### PR DESCRIPTION
**Pull request by rshriram**
Follow up from @rodaine's route-local config framework. Enabling per-route/per-vhost config for fault filters. (https://github.com/envoyproxy/envoy/pull/3045).

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

**Comment by mattklein123** [comment]
@rodaine can you own first review on this.

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 1]
```diff
@@ -71,6 +69,29 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
...
+
+    if (proto_config) {
+      const envoy::config::filter::http::fault::v2::HTTPFault per_filter_config =
+          dynamic_cast<const envoy::config::filter::http::fault::v2::HTTPFault&>(*proto_config);
+
+      config_.reset(new FaultFilterConfig(per_filter_config, config_->runtime(),
+                                          config_->statsPrefix(), config_->scope()));
+    }
+  }
+
```
Comment (rodaine) [inline source/extensions/filters/http/fault/fault_filter.cc]: i think you can get away with doing this in the `setDecoderFilterCallbacks` method, no? Not sure if the route is known by then...but it seems to work from my tests for the buffer filter.

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 2]
```diff
@@ -71,6 +69,29 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
...
+  // Its okay for the moment, since the only use case of faults with runtimes is
+  // by folks using filter level configuration in the fault filter (mainly Lyft folks).
+  if (callbacks_->route() && callbacks_->route()->routeEntry()) {
+    auto proto_config = callbacks_->route()->routeEntry()->perFilterConfig(
+        Extensions::HttpFilters::HttpFilterNames::get().FAULT);
+    if (!proto_config) {
+      // See if the virtual host has any fault filter config
+      proto_config = callbacks_->route()->routeEntry()->virtualHost().perFilterConfig(
+          Extensions::HttpFilters::HttpFilterNames::get().FAULT);
+    }
```
Comment (rodaine) [inline source/extensions/filters/http/fault/fault_filter.cc]: Can roll this up into a ternary. Something like: 
```c++
const std::string name = Extensions::HttpFilters::HttpFilterNames::get().FAULT;
const auto* route_entry = callbacks_->route()->routeEntry();

const auto* proto_config = route_entry->perFilterConfig(name) 
    ?: route_entry->virtualHost().perFilterConfig(name);

// ...
```

**Comment by rshriram** [comment]
@mattklein123 / anybody else, could you please review this PR? sorry, trying to get this in before the next istio RC release.

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 3]
```diff
@@ -36,10 +38,6 @@ FaultFilterConfig::FaultFilterConfig(const envoy::config::filter::http::fault::v
     : runtime_(runtime), stats_(generateStats(stats_prefix, scope)), stats_prefix_(stats_prefix),
       scope_(scope) {
 
-  if (!fault.has_abort() && !fault.has_delay()) {
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: Why was this removed?

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 4]
```diff
@@ -71,6 +69,27 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
 // if we inject a delay, then we will inject the abort in the delay timer
 // callback.
 Http::FilterHeadersStatus FaultFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  // Route-level configuration overrides filter-level configuration
+  // TODO (rshriram): We should not be using Runtimes when reading from route-level
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: s/Runtimes/runtime

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 5]
```diff
@@ -71,6 +69,27 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
...
+  // faults until we parameterize the runtime keys for faults based on the route name.
+  // Its okay for the moment, since the only use case of faults with runtimes is
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: s/Its/It's

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 6]
```diff
@@ -71,6 +69,27 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
...
+  // by folks using filter level configuration in the fault filter (mainly Lyft folks).
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: remove "(mainly Lyft folks)"

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 7]
```diff
@@ -71,6 +69,27 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
...
+    const auto* route_entry = callbacks_->route()->routeEntry();
+
+    const auto* proto_config =
+        route_entry->perFilterConfig(name) ?: route_entry->virtualHost().perFilterConfig(name);
+
+    if (proto_config) {
+      const envoy::config::filter::http::fault::v2::HTTPFault per_filter_config =
+          dynamic_cast<const envoy::config::filter::http::fault::v2::HTTPFault&>(*proto_config);
+
+      config_.reset(new FaultFilterConfig(per_filter_config, config_->runtime(),
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: This is fine, but there would be a perf benefit for using the update callback mechanism that is being added. I would add a TODO to improve this later.

File: source/extensions/filters/http/fault/fault_filter.cc (hunk 3 above)
Comment (rshriram) [reply source/extensions/filters/http/fault/fault_filter.cc]: This is to support a fully empty fault filter config -- which essentially should do nothing. This will allow keeping the fault filter always on and shipping filter configs via RDS

**Comment by rshriram** [comment]
done..

File: source/extensions/filters/http/fault/fault_filter.cc (hunk 3 above)
Comment (mattklein123) [reply source/extensions/filters/http/fault/fault_filter.cc]: OK thanks

**Comment by mattklein123** [comment]
Holding on merging until we resolve https://github.com/envoyproxy/envoy/pull/3128#pullrequestreview-113377839. cc @ggreenway @rodaine

**Comment by mattklein123** [comment]
@rshriram can you merge master and fix this up? Then we can get this in.

**Comment by rshriram** [comment]
@mattklein123 updated

File: source/extensions/filters/http/fault/config.cc [hunk 8]
```diff
@@ -42,6 +42,15 @@ FaultFilterFactory::createFilterFactoryFromProto(const Protobuf::Message& proto_
       stats_prefix, context);
 }
 
+Router::RouteSpecificFilterConfigConstSharedPtr
+FaultFilterFactory::createRouteSpecificFilterConfig(const ProtobufWkt::Struct& struct_config) {
+  envoy::config::filter::http::fault::v2::HTTPFault proto_config;
+  MessageUtil::jsonConvert(struct_config, proto_config);
+  Router::RouteSpecificFilterConfigConstSharedPtr rconfig;
+  rconfig.reset(new Fault::FaultSettings(proto_config));
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/config.cc]: nit: `return std::make_shared<...>(...)`

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 9]
```diff
@@ -71,6 +71,20 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
 // if we inject a delay, then we will inject the abort in the delay timer
 // callback.
 Http::FilterHeadersStatus FaultFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  // Route-level configuration overrides filter-level configuration
+  // NOTE: We should not use runtime when reading from route-level
+  // faults. In other words, runtime is supported only when faults are
+  // configured at the filter level.
+  if (callbacks_->route() && callbacks_->route()->routeEntry()) {
+    const std::string name = Extensions::HttpFilters::HttpFilterNames::get().FAULT;
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: nit: `const std::string&`

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 10]
```diff
@@ -71,6 +71,20 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
...
+    const auto* route_entry = callbacks_->route()->routeEntry();
+
+    const FaultSettings* per_route_settings =
+        route_entry->perFilterConfigTyped<FaultSettings>(name)
+            ?: route_entry->virtualHost().perFilterConfigTyped<FaultSettings>(name);
+    config_->updateFaultSettings(per_route_settings);
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: Aren't you effectively updating a global shared pointer here? I don't think this is right? Can you make sure `FaultFilterConfigSharedPtr` is actually `FaultFilterConfigConstSharedPtr` to catch this?

Comment (rshriram) [reply source/extensions/filters/http/fault/fault_filter.cc]: ack.. good catch..

Comment (rshriram) [reply source/extensions/filters/http/fault/fault_filter.cc]: now that I think about it, the reason I did this was because I wanted to avoid allocation in the request path. I couldn't create an identical config object for RouteSpecific.. because I don't have access to runtime/stats/scope etc in that API Call.

Comment (mattklein123) [reply source/extensions/filters/http/fault/fault_filter.cc]: Hmm. Then this is the same problem as Lua. We need to fix this now I guess and be able to pass the full FactoryContext into the factory function. Can you take a look at that?

File: source/extensions/filters/http/fault/fault_filter.h [hunk 11]
```diff
@@ -43,35 +70,20 @@ class FaultFilterConfig {
...
-  std::unordered_set<std::string> downstream_nodes_{}; // Inject failures for specific downstream
-                                                       // nodes. If not set then inject for all.
   Runtime::Loader& runtime_;
   FaultFilterStats stats_;
   const std::string stats_prefix_;
   Stats::Scope& scope_;
+  const FaultSettings settings_;
 };
 
 typedef std::shared_ptr<FaultFilterConfig> FaultFilterConfigSharedPtr;
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.h]: Can you change this typedef to `std::shared_ptr<const FaultFilterConfig>`

**Comment by mattklein123** [comment]
@rshriram check OSX failure, it's real compiler error.

**Comment by rshriram** [comment]
so, I cant really do `const filterconfig..` because the filter config has stats as well, which get incremented whenever faults are injected..

**Comment by rshriram** [comment]
Tests passing finally.
